crates/ra_ide_api/src/hover.rs

```diff
@@ -162,55 +163,56 @@ fn hover_text_from_name_kind(
-    let parse = db.parse(position.file_id);
-    let file = parse.tree();
+    let file = db.parse_or_expand(position.file_id.into())?;
+    let token = file.token_at_offset(position.offset).filter(|it| !it.kind().is_trivia()).next()?;
+    let token = descend_into_macros(db, position.file_id, token);
 
     let mut res = HoverResult::new();
 
-    let mut range = if let Some(name_ref) =
-        find_node_at_offset::<ast::NameRef>(file.syntax(), position.offset)
-    {
-        let mut no_fallback = false;
-        if let Some(name_kind) =
-            classify_name_ref(db, Source::new(position.file_id.into(), &name_ref)).map(|d| d.kind)
-        {
-            res.extend(hover_text_from_name_kind(db, name_kind, &mut no_fallback))
-        }
+    let mut range = match_ast! {
+        match (token.ast.parent()) {
+            ast::NameRef(name_ref) => {
+                let mut no_fallback = false;
+                if let Some(name_kind) =
+                    classify_name_ref(db, token.with_ast(&name_ref)).map(|d| d.kind)
+                {
+                    res.extend(hover_text_from_name_kind(db, name_kind, &mut no_fallback))
+                }
 
-        if res.is_empty() && !no_fallback {
-            // Fallback index based approach:
-            let symbols = crate::symbol_index::index_resolve(db, &name_ref);
-            for sym in symbols {
-                let docs = docs_from_symbol(db, &sym);
-                let desc = description_from_symbol(db, &sym);
-                res.extend(hover_text(docs, desc));
-            }
-        }
+                if res.is_empty() && !no_fallback {
+                    // Fallback index based approach:
+                    let symbols = crate::symbol_index::index_resolve(db, &name_ref);
+                    for sym in symbols {
+                        let docs = docs_from_symbol(db, &sym);
+                        let desc = description_from_symbol(db, &sym);
+                        res.extend(hover_text(docs, desc));
+                    }
+                }
 
-        if !res.is_empty() {
-            Some(name_ref.syntax().text_range())
-        } else {
-            None
-        }
-    } else if let Some(name) = find_node_at_offset::<ast::Name>(file.syntax(), position.offset) {
-        if let Some(name_kind) =
-            classify_name(db, Source::new(position.file_id.into(), &name)).map(|d| d.kind)
-        {
-            let mut _b: bool = true;
-            res.extend(hover_text_from_name_kind(db, name_kind, &mut _b));
-        }
+                if !res.is_empty() {
+                    Some(name_ref.syntax().text_range())
+                } else {
+                    None
+                }
+            },
+            ast::Name(name) => {
+                if let Some(name_kind) = classify_name(db, token.with_ast(&name)).map(|d| d.kind) {
+                    let mut _b: bool = true;
```

Comment:
Nit: `&mut true`

Reply:
Oh cool I didn't know you could do that.